### ISSUE TITLE
PTL gradient accumulation + drop_last bug

### ISSUE THREAD

**Issue by isaacrob** [roboflow/rf-detr#916]
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

This PTL bug https://github.com/Lightning-AI/pytorch-lightning/issues/19987 appears to be impacting the new PTL RF-DETR implementation.

We use gradient accumulation to allow us to run at an effective batch size larger than can fit in memory at once. So for example we could have an effective batch size of 16 by setting bs = 16 and ga = 1, or bs = 2 and ga = 8. In the original implementation, the data loader would fill batches of effective_batch_size size, and would use drop_last to avoid filling a batch where there wasn't enough data to make a similarly scaled step. It would then split the batch into ga subsets and step on those.

In the new PTL implementation, the data module is grabbing local bs chunks and using drop last. If the last batch only has 8 images instead of 16, it'll do ga 4 times instead of 8 and step, making the last step effectively a step over bs 8 instead of bs 16, which is the thing that drop_last is meant to solve and can cause instability immediately before the validation runs. 

Here are the three lines that together produce the bug:

  1. module_data.py:131-135 — DataLoader uses local batch_size with `drop_last=True`
   This drops the tail at local-batch granularity. If the dataset has 118,287 images and bs=2, this yields 59,143 local batches. With ga=8, that's 7,393 full accumulation groups + 7 leftover batches.
		```py
		return DataLoader(
		    dataset,
		    batch_size=batch_size,       # line 133: local bs, NOT effective bs
		    shuffle=True,
		    drop_last=True,              # line 135: drops incomplete LOCAL batch only
		    ...
		)
		```

  2. trainer.py:233 — grad_accum_steps wired to PTL
  `"accumulate_grad_batches": tc.grad_accum_steps,`
  PTL consumes these 59,143 local batches in groups of 8. After 7,393 full groups (59,144 batches), there are 7 remaining. PTL fires the optimizer on these 7 anyway — a partial accumulation window.

  3. module_model.py:154 — loss unconditionally divided by full accumulate_grad_batches
  `loss_scaled = loss / self.trainer.accumulate_grad_batches`
  Each of those 7 sub-batches divides its loss by 8 (the configured value), not by 7 (the actual accumulation count). So the final optimizer step receives 7/8 of the gradient it should — it's systematically under-scaled.

### Environment

RF-DETR 1.6.3

### Minimal Reproducible Example
```py
  from rfdetr import RFDETRNano

  model = RFDETRNano()

  # 10 images, bs=3, ga=2 → 3 local batches after drop_last → 1 full group + 1 partial
  # The partial group (1 sub-batch) still divides loss by ga=2 → half-strength update
  model.train(
      dataset_dir="path/to/small-dataset",  # any dataset with ~10 images
      batch_size=3,
      grad_accum_steps=2,
      epochs=1,
  )
```
### Additional

@Borda I'm not sure the best way to solve this within PTL land

### Are you willing to submit a PR?

- [x] Yes, I'd like to help by submitting a PR!

**Comment by Borda** [comment]
Hi @isaacrob — your analysis is spot-on on all three counts. The `drop_last=True` at local-batch granularity, the static `accumulate_grad_batches` divisor in the partial window, and the upstream PTL issue (pytorch-lightning#19987) being stalled — I've confirmed each of these.

I'm going to fix this with **dataset padding rather than truncation.** The training dataset will be wrapped so its length is the next multiple of `effective_batch_size = batch_size × grad_accum_steps × world_size`. Padding samples are drawn randomly from the original dataset, but because RF-DETR uses online augmentation, each padded sample gets a fresh random augmentation at `__getitem__` time — so it behaves like a new training example rather than a true duplicate. This means zero data is dropped, which is the right call for a training pipeline that generates novel views on the fly.

The wrapper will live in `module_data.py` and is explicitly scoped to be easy to lift once pytorch-lightning#19987 ships a native fix.

If you'd like to submit a PR for this, I'd welcome it — the scope is fairly self-contained. Otherwise I'll pick it up. Either way, happy to talk through any details before you start.

**Comment by isaacrob** [comment]
Why pad as opposed to drop? What benefit is expected?

**Comment by abhiagwl4262** [comment]
@Borda this might be causing multi-gpu training fails ?

**Comment by Borda** [comment]
> Why pad as opposed to drop? What benefit is expected?

For a small datasets it is more important to use all avalaibel data and with online augmentation the data will be always a bit different...

**Comment by Borda** [comment]
> Borda this might be causing multi-gpu training fails ?

Why do you think so?

**Comment by abhiagwl4262** [comment]
> > Borda this might be causing multi-gpu training fails ?
> 
> Why do you think so?

Because you told that the issue might occur because of different no of samples getting processed on different GPUs which can happen if we do not drop the last batch

**Comment by Borda** [comment]
> > > Borda this might be causing multi-gpu training fails ?
> > 
> > 
> > Why do you think so?
> 
> Because you told that the issue might occur because of different no of samples getting processed on different GPUs which can happen if we do not drop the last batch

right that would happen on any device, even on CPU, so yes, it could also be multi-GPU, but the issue is the same...
